### PR TITLE
Testing new apply interface for Flux.Chain

### PR DESCRIPTION
**Pull request by mkschleg**
New Feature:

Implemented new `apply` functionality which returns a newly constructed chain as well as the output of applying the input to the current chain. This is in relation to the recent conversation for updating the [recurrent network interface](https://github.com/FluxML/Flux.jl/issues/1678#issuecomment-1440987414). The apply was described by @ToucheSir.

Most of the code is a modification of the `_applychain` functionality from Flux.jl. The general direction is to likely interface with Accessors.jl. Some minor tests were added, but we will want to include many of the layer types as we expand the feature.

### PR Checklist

- [x] Tests are added
- [ ] Documentation, if applicable

### CODE REVIEW

File: src/chain.jl [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+end
+
+# Generated function returns a tuple of args and the last output of the network.
+@generated function apply(layers::Tuple{Vararg{<:Any,N}}, x) where {N}
+  x_symbols = vcat(:x, [gensym() for _ in 1:N])
+  l_symbols = [gensym() for _ in 1:N]
+  calls = [:(($(l_symbols[i]), $(x_symbols[i+1])) = apply(layers[$i], $(x_symbols[i]))) for i in 1:N]
+  push!(calls, :(return tuple($(l_symbols...)), $(x_symbols[end])))
+  Expr(:block, calls...)
+end
```
Comment (ToucheSir) [inline src/chain.jl:33]: I would recommend making these internal functions instead of additional methods on `apply`.

File: src/chain.jl [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+apply(layers::NamedTuple{NMS, TPS}, x) where {NMS, TPS} = begin
+  layers, out = apply(Tuple(layers), x)
+  NamedTuple{NMS}(layers), out
+end
+
+function apply(layers::AbstractVector, x)  # type-unstable path, helps compile times
+  new_layers = typeof(layers)(undef, length(layers))
+  for (idx, f) in enumerate(layers)
+    new_layers[idx], x = apply(f, x)
+  end
```
Comment (ToucheSir) [inline src/chain.jl:22]: This won't be differentiable, unfortunately. I don't know if there are any non-hacky tricks to make it work. We might have to define our own `scan` function or something like it.

Comment (mkschleg) [reply src/chain.jl]: Ok. Yeah, I tried a few other options, but I think a `_scan` function with a hand written rrule is likely what we need. I'm not too adept at writing rrules yet, but will work on it. In the mean time I've added some tests for checking gradient consistency and the scan function with a placeholder for the rrule and skip the gradient test for the vector version.

File: src/chain.jl (hunk 1 above)
Comment (mkschleg) [reply src/chain.jl]: Done. Added an `_apply` function to work on the set of layers, and then `_apply_to_layer` which will be the layer by layer interface. Seemed cleaner to separate the two.

File: src/chain.jl [hunk 3]
```diff
@@ -0,0 +1,49 @@
+
+import Flux: ChainRulesCore
+# Some experiments with chain to start removing the need for recur to be mutable.
+# As per the conversation in the recurrent network rework issue.
+
+# Main difference between this and the _applychain function is we return a new chain
+# with the internal state modified as well as the output of applying x to the chain.
+apply(chain::Flux.Chain, x) = begin
```
Comment (ToucheSir) [inline src/chain.jl:8]:
```suggestion
function apply(chain::Flux.Chain, x)
```

File: src/chain.jl [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+  layers, out = _apply(chain.layers, x)
+  Flux.Chain(layers), out
+end
+
+_apply(layers::NamedTuple{NMS, TPS}, x) where {NMS, TPS} = begin
```
Comment (ToucheSir) [inline src/chain.jl:13]:
```suggestion
function _apply(layers::NamedTuple{NMS, TPS}, x) where {NMS, TPS}
```

File: src/chain.jl [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+  for (idx, f) in enumerate(layers)
+    new_layers[idx], x = _apply_to_layer(f, x)
+  end
+  new_layers, x
+end
+
+function ChainRulesCore.rrule(::typeof(_scan), layers, x)
+  function _scan_pullback(dy)
+    throw("_scan Pullback not implemented")
+    return (ChainRulesCore.NoTangent(), ChainRulesCore.NoTangent())
```
Comment (ToucheSir) [inline src/chain.jl:29]:
```suggestion
    error("_scan Pullback not implemented")
```

Comment (ToucheSir) [reply src/chain.jl:29]: This should return `ChainRulesCore.NotImplemented`, but currently Zygote (incorrectly) ignores that.

File: test/chain.jl [hunk 6]
```diff
@@ -0,0 +1,85 @@
+
+import Flux, Fluxperimental
+
```
Comment (ToucheSir) [inline test/chain.jl:3]:
```suggestion
```
This file could use a general whitespace and dead comment removal pass, I think.

Comment (mkschleg) [reply test/chain.jl]: Just added new commit which goes through and cleans up the test file.

**Comment by mkschleg** [comment]
I added the pullback for the vector chain. It likely doesn't work on GPUs yet, and may not be implemented in the most efficient way possible. But it passes the simple test comparing to the other interface.

**Comment by mkschleg** [comment]
I also decided to unify _apply with _apply_to_layer because I initially forgot about chains which can be composed of other chains through the parallel and other interfaces.

**Comment by mkschleg** [comment]
The non-mutating recur is relatively straightforward. One snag is we can no longer us the simple forward pass for time series data of the old mutating version (i.e. `[chain(x) for x in xs]`). Instead we need another scan like function. In Equinox, I believe they use jax.vmap to handle this (which maps over the time access of their inputs). I could see two options.

1. Implement a very simple map function ontop of chain which sends an input through the entire model before moving to the next timestep. Likely could use `accumulate` as a simple starting point.
2. Send the vector of time inputs through the chain layer by layer (kinda similar to how we handle blocked time right now in Flux.Recur). Only the stateful layers would have to be treated specially in this case. 

I'm not as in on the overarching plans for Flux, but I could see 2 offering more options for optimization.

**Comment by mkschleg** [comment]
Also, implementing the non-mutating recur for this interface requires us to make a decision about how to handle time-series inputs for a chain. This is because `[chain(x) for x in xs]` will no longer work (for hopefully clear reasons). Instead, we will need a custom map function (which maybe `_apply` can dispatch on?). I could see this kind of function going two ways:

1. We handle this at the model level, and run the forward pass for the entire model on each time-step of the input.
2. We handle this at the layer level, and run the forward pass for the entire input for each layer sequentially.

I could see 2 offering more optimization options. But I'm not sure how well it will interact with the rest of Flux (especially future planned stuff).

**Comment by mkschleg** [comment]
@ToucheSir For sure. And yes it is, but the 3d api was never meant to be fully featured from what I remember when we were working on it. I'll put up another PR for the non-mutating recur when I make some more progress.
